include/graaflib/algorithm/topological_sorting.tpp

```diff
@@ -0,0 +1,60 @@
+#pragma once
+
+#include <graaflib/algorithm/cycle_detection.h>
+#include <graaflib/algorithm/topological_sorting.h>
+
+#include <vector>
+#include <unordered_map>
+
+namespace graaf::algorithm {
+
+namespace detail {
+
+enum class vertex_status { UNVISITED, VISITED};
+
+// DFS topological sort
+template <typename V, typename E>
+void do_dfs_topological_sort(
+    const graph<V, E, graph_type::DIRECTED>& graph,
+    vertex_id_t start_vertex,
+    std::unordered_map<vertex_id_t, vertex_status>& processed_vertices,
+    std::vector<vertex_id_t>& sorted_vertices) {
+    
+    processed_vertices[start_vertex] = vertex_status::VISITED;
+    for (const auto& next_vertex : graph.get_neighbors(start_vertex)) {
+        if (processed_vertices[next_vertex] == vertex_status::UNVISITED) {
+            do_dfs_topological_sort(graph, next_vertex, processed_vertices,
+                sorted_vertices);
+        }
+    }
+
+        sorted_vertices.push_back(start_vertex);
+    }
+
+}; // namespace detail
+
+template <typename V, typename E>
+std::vector<vertex_id_t> topological_sort(
+    const graph<V, E, graph_type::DIRECTED>& graph) {
+    
+    // Graph should be acyclic
+    if (dfs_cycle_detection(graph)) {
+        return {};    
+    }
+
+    using enum detail::vertex_status;
+    std::vector<vertex_id_t> sorted_vertices{};
+    sorted_vertices.reserve(graph.vertex_count());
+    std::unordered_map<vertex_id_t, detail::vertex_status> processed_vertices{};
```

Comment:
do we really _need_ the `vertex_status` enum? Wouldn't a container that holds the visited vertices suffice? Also, it looks like we never set (i.e. initialize) something to `UNVISITED`

Reply:
Added for more readability. Let me check if the built-in container will do the job. But what about if the user doesn't want to fill the inner container?

Reply:
Looking at how the container is used currently, the "initialization" is done by calling `operator[]`. Since this operator default constructs values for missing keys, they are initialized to the first enum value (`UNVISITED`).

Since this data structure is not exposed to the user of the algorithm, we can treat it as an implementation detail and freely change the underlying container type. So no need to provide multiple implementations.

What do you think about using an `std::unordered_set` here? When the vertex is visited you can insert into it and the rest of the code could do something like `processed_vertices.contains(current_vertex_id)`.